control/frdata.py

```diff
@@ -342,110 +336,107 @@ def __pow__(self, other):
             else:
-                for i in range(self.outputs):
-                    for j in range(self.inputs):
-                        frraw = splev(omega, self.ifunc[i, j], der=0)
-                        out[i, j] = frraw[0] + 1.0j * frraw[1]
-
+                out = self.fresp[:, :, elements]
+        else:
+            out = empty((self.outputs, self.inputs, len(omega_array)), 
+                         dtype=complex)
+            for i in range(self.outputs):
+                for j in range(self.inputs):
+                    for k, w in enumerate(omega_array):
+                        frraw = splev(w, self.ifunc[i, j], der=0)
+                        out[i, j, k] = frraw[0] + 1.0j * frraw[1]
+        if not hasattr(omega, '__len__'):
+            # omega is a scalar, squeeze down array along last dim
+            out = np.squeeze(out, axis=2)
+        if squeeze and self.issiso():
+            out = out[0][0]
         return out
 
-    # Method for generating the frequency response of the system
-    def freqresp(self, omega):
-        """Evaluate the frequency response at a list of angular frequencies.
+    def __call__(self, s, squeeze=True):
+        """Evaluate system's transfer function at complex frequencies.
+        
+        Returns the complex frequency response `sys(s)`.
 
-        Reports the value of the magnitude, phase, and angular frequency of
-        the requency response evaluated at omega, where omega is a list of
-        angular frequencies, and is a sorted version of the input omega.
+        To evaluate at a frequency omega in radians per second, enter 
+        x = omega*1j or use FRD.eval(omega)
 
         Parameters
         ----------
-        omega : array_like
-            A list of frequencies in radians/sec at which the system should be
-            evaluated. The list can be either a python list or a numpy array
-            and will be sorted before evaluation.
-
+        s: complex scalar or array_like 
+            Complex frequencies 
+        squeeze: bool, optional (default=True)
+            If True and sys is single input single output (SISO), returns a 
+            1D array or scalar depending on the length of x. 
+            
         Returns
         -------
-        mag : (self.outputs, self.inputs, len(omega)) ndarray
-            The magnitude (absolute value, not dB or log10) of the system
-            frequency response.
-        phase : (self.outputs, self.inputs, len(omega)) ndarray
-            The wrapped phase in radians of the system frequency response.
-        omega : ndarray or list or tuple
-            The list of sorted frequencies at which the response was
-            evaluated.
-        """
-        # Preallocate outputs.
-        numfreq = len(omega)
-        mag = empty((self.outputs, self.inputs, numfreq))
-        phase = empty((self.outputs, self.inputs, numfreq))
+        fresp : (num_outputs, num_inputs, len(x)) or len(x) complex ndarray
```

Comment:
```suggestion
        fresp : (num_outputs, num_inputs, len(x)) ndarray or (len(x), ) complex ndarray
```

Reply:
This pushes the length of the row past 80 characters. ok to make an exception in that case to

`fresp : (self.outputs, self.inputs, len(x)) or len(x) complex ndarray` (78 characters including tabs) which I think is still understandable?

Not sure what the alternative is. maybe:
`        fresp : (self.outputs, self.inputs, len(x)) complex ndarray or 
                (len(x),) complex ndarray`

Reply:
hey @bnavigator maybe we can resolve these comments since we're here. Do you think what ive suggested above, to overcome line exceeding 80 chars, makse sense fr this and the ines below?

Reply:
Sorry this was hidden in the notifications.

Let me check with spyder how it renders real quick...

Reply:
```
        Returns
        -------
        fresp : (self.outputs, self.inputs, len(s)) or (len(s),) complex ndarray
            The frequency response of the system. Array is ``len(s)`` if and
            only if system is SISO and ``squeeze=True``.
        fresp : (self.outputs, self.inputs, len(s)) or len(s) complex ndarray
            The frequency response of the system. Array is ``len(s)`` if and
            only if system is SISO and ``squeeze=True``.
        fresp : (self.outputs, self.inputs, len(x)) complex ndarray or
                (len(x),) complex ndarray
            The frequency response of the system. Array is len(x) if and only
            if system is SISO and squeeze=True.
```

![Screenshot_20210107_004951](https://user-images.githubusercontent.com/4623504/103833859-48a9f300-5082-11eb-9df9-ae0592965e02.png)

Doesn't look good

Reply:
nope. maybe best to leave > 80 chars & be precise

Reply:
How about?

```
        Returns the complex frequency response `sys(s)` of system `sys` with
        `m = sys.inputs` number of inputs and `p = sys.outputs` number of
        outputs.

        To evaluate at a frequency omega in radians per second, enter
        ``s = omega * 1j`` or use ``sys.eval(omega)``

        Parameters
        ----------
        s: complex scalar or array_like
            Complex frequencies
        squeeze: bool, optional (default=True)
            If True and `sys` is single input single output (SISO), i.e. `m=1`,
            `p=1`, return a 1D array or scalar depending on the length of `s`.

        Returns
        -------
        fresp : (p, m, len(s)) or (len(s),) complex ndarray or complex scalar
            The frequency response of the system. Array is ``(len(s), )`` if
            and only if system is SISO and ``squeeze=True``.
```

![Screenshot_20210107_010323](https://user-images.githubusercontent.com/4623504/103834617-33ce5f00-5084-11eb-9480-bad46bae9b99.png)

Reply:
looks good. except: if len(s) is 1 and SISO and squeeze, it still returns an ndarray of length (1,). 

is that technically a scalar?

Reply:
No it's not. Then the current text for the squeeze parameter is wrong.

Reply:
... or you have to adjust the code so that the doc becomes true :wink: